### PR TITLE
C.65 Self-move - require only valid but unspecified

### PR DESCRIPTION
**Pull request by bgloyer**
Cleaned up [C.65: Make move assignment safe for self-assignment](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-move-self)
based on the discussions in #1892  #1649 #1606.

### CODE REVIEW

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -6215,49 +6214,44 @@ Unless there is an exceptionally strong reason not to, make `x = std::move(y); y
...
+
+    class X {
     public:
-        Foo& operator=(Foo&& a);
+        X();
+        X& operator=(X&& a) noexcept;
         // ...
+        ~X() { delete owning_ptr; }
+    private:
+        T* owning_ptr;  // bad (See R.20) but used in the example because
```
Comment (bgloyer) [inline CppCoreGuidelines.md:6239]: I couldn't think of a simple example that required a manual move-assignment that didn't go against something in the guidelines

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -6215,49 +6214,44 @@ Unless there is an exceptionally strong reason not to, make `x = std::move(y); y
...
+                        // it requires a manual move-assignment 
     };
 
-    Foo& Foo::operator=(Foo&& a) noexcept  // OK, but there is a cost
+    X& X::operator=(X&& a) noexcept
```
Comment (bgloyer) [inline CppCoreGuidelines.md:6243]: Used 'X' instead of 'Foo' to match C.64

**Comment by cubbimew** [comment]
This seems more radical than #1606:
* current C.65 requires that self-move is a no-op and insists on not doing the STL thing (self-destructive self-move)
* #1606 points out that it is incompatible with rule of zero for classes that have members of STL types but can be worked around with copy/swap,
* this PR flips C.65 around to do the STL thing and not require no-op

As I recall, response to #1606 stalled because there was a draft paper on that topic.

**Comment by bgloyer** [comment]
Can we use 'valid-but-uspecified' until it is offically settled?  I don't think it is practical to have the rule stronger than what the stl types do.  Any type with an stl memeber that uses the rule of zero or `= default` would violate this rule.  Or maybe have two rules, one that says 'valid-but-unspecified' and a stronger one for 'nop'?  Then developers could choose which is appropriate for them?  Setting 'valid-but-uspecified' as a minimum is important so that a type will work with `swap`.

**Comment by MikeGitb** [comment]
I like the Idea to stick with the stl behavior for now,  but would love to see more data about the need for stronger rules.  Has anyone had anekdotal or regular encounters with a self-move outside of self swap (where you don't care about the value)? What were the consequences for this?

**Comment by hsutter** [comment]
Editors call: This would invert the guidance and we agree with the current guidance -- self-move should be a  no-op, even though the standard library does not currently/yet require that.

**Pull request closed without merging**